### PR TITLE
feat: add control api for discovery module

### PR DESCRIPTION
**Pull request by yongboy**
### What this PR does / why we need it:

Just implement the `_M.dump_data()` function for discovery module, then call the control api for dumping running details online.

The dump api uri template:

```shell
GET /v1/discovery/{discovery_type}/dump
```

eg:

```
curl http://127.0.0.1:9090/v1/discovery/eureka/dump
```

### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [x] Have you modified the corresponding document?
* [x] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**

### CODE REVIEW

**Comment by spacewander** [comment]
The conflict needs to be solved so that the CI can run.

File: docs/en/latest/discovery.md [hunk 1]
```diff
@@ -66,7 +67,9 @@ It is very easy for APISIX to extend the discovery client, the basic steps are a
 
 2. Implement the `_M. init_worker()` function for initialization and the `_M. nodes(service_name)` function for obtaining the list of service instance nodes;
 
-3. Convert the registry data into data in APISIX;
+3. If you need the discovery module to export the debugging information online,  implement the `_M. dump_data()` function;
```
Comment (spacewander) [inline docs/en/latest/discovery.md:70]: Style: extra space before `dump_data`.

File: apisix/control/router.lua [hunk 2]
```diff
@@ -66,6 +68,38 @@ function fetch_control_api_router()
         end
     end
 
+    local discovery_type = require("apisix.core.config_local").local_conf().discovery
+    if discovery_type then
+        local discovery = require("apisix.discovery.init").discovery
+        local dump_apis = {}
+        for key, _ in pairs(discovery_type) do
```
Comment (spacewander) [inline apisix/control/router.lua:75]: It seems we can use `for key, dis_mod in pairs(discovery) do` directly?

File: t/control/discovery.t [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+__DATA__
+
+=== TEST 1: test consul_kv dump_data api
+--- yaml_config eval: $::yaml_config
+--- request
+GET /v1/discovery/consul_kv/dump
+--- error_code: 200
+--- response_body_unlike
+^{}$
```
Comment (spacewander) [inline t/control/discovery.t:64]: Can we check part of the content in this way: https://github.com/apache/apisix/blob/8e3fd226714d0a733b87aac33f64fe97e77d3785/t/control/schema.t#L45

File: t/control/discovery.t [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+
+
+=== TEST 2: test eureka dump_data api
+--- yaml_config eval: $::yaml_config
+--- request
+GET /v1/discovery/eureka/dump
+--- error_code: 200
+--- response_body_unlike
+^{}$
```
Comment (spacewander) [inline t/control/discovery.t:74]: ditto

**Comment by spacewander** [comment]
The CI fails because of https://github.com/apache/apisix/pull/3747

File: docs/en/latest/discovery.md (hunk 1 above)
Comment (yongboy) [reply docs/en/latest/discovery.md:70]: is `_M. dump_data()` -> `_M.dump_data()` ?

File: docs/en/latest/discovery.md [hunk 5]
```diff
@@ -92,6 +95,11 @@ Then implement the `_M.init_worker()` function for initialization and the `_M.no
   end
 
 
+  function _M.dump_data()
+      ... ...
```
Comment (tokers) [inline docs/en/latest/discovery.md:99]: I think an example return data shoul be written here.

File: docs/en/latest/discovery.md (hunk 1 above)
Comment (spacewander) [reply docs/en/latest/discovery.md:70]: Yes

File: docs/en/latest/discovery.md (hunk 5 above)
Comment (yongboy) [reply docs/en/latest/discovery.md:99]: for example ?
```lua
  function _M.dump_data()
      return {config = your_config, services = your_services, other = ... } 
  end
```

File: t/control/discovery.t (hunk 3 above)
Comment (yongboy) [reply t/control/discovery.t:64]: Y

File: t/control/discovery.t (hunk 4 above)
Comment (yongboy) [reply t/control/discovery.t:74]: Y

File: apisix/control/router.lua (hunk 2 above)
Comment (yongboy) [reply apisix/control/router.lua:75]: if use `for key, dis_mod in pairs(discovery) do`, the `dis_mod` is `function` type, not a table object.

So, we can't do it.

**Review by spacewander — APPROVED** [review]

**Review by tokers — APPROVED** [review]

**Pull request merged**